Separate autopilot multirotor from autopilot wing - dummy files for wing

limonspb opened the pull request:
This PR has no code changes to the existing multirotor autopilot/gps rescue/poshold/althold.
it is a second attempt to separate autopilot/gps rescue for wings from multirotor.
Previous one: https://github.com/betaflight/betaflight/pull/14011 (closing now)
**This PR has NO code changes**, only covers the existing multirotor code with `#ifNdef USE_WING` and some files renaming as per discussion with @ledvinap here:
https://github.com/betaflight/betaflight/pull/14011#issuecomment-2464548224
This PR will allow me to finish GPS rescue code for wings separate from the quads/multirotor improvement and continue with althold/poshold and autolaunch.

`autopilot.c/h` renamed to `autopilot_multirotor.c/h` and added `autopilot_wing.c/h` with dummy functions to be filled in future PR.
`autopilot.c/h` will only include `autopilot_multirotor.h` and `autopilot_wing.h`.
But in future a common code can be moved to `autopilot.h`.

All client code should just include `autopilot.h`, just like it is currently now in master, no changes here.

Same was done for `gps_rescue.c/h`, `alt_hold.c/h` and `pos_hold.c/h`.

GPS rescue CMS menu also has two separate files now: `cms_menu_gps_rescue_multirotor.c` and `cms_menu_gps_rescue_wing.c`.
@haslinghuis @ctzsnooze @KarateBrot  @ledvinap

KarateBrot commented:
I feel like this goes into the wrong direction. We can probably use abstraction and polymorphism to create a shared auropulot interface for both modes (multirotor and wing).

ctzsnooze reviewed (approved):
Approve so long as existing functionality is not adversely affected.

haslinghuis commented on src/main/pg/gps_rescue_multirotor.h:
```diff
@@ -0,0 +1,54 @@
+/*
```
Wrong license file

haslinghuis commented on src/main/pg/gps_rescue_wing.c:
```diff
@@ -0,0 +1,43 @@
+/*
```
wrong license file

haslinghuis commented on src/main/pg/gps_rescue_wing.h:
```diff
@@ -0,0 +1,36 @@
+/*
```
Wrong license file

haslinghuis commented on src/main/cms/cms_menu_gps_rescue_wing.c:
```diff
@@ -0,0 +1,117 @@
+/*
```
wrong license file

haslinghuis reviewed (dismissed):
Like @KarateBrot see a lot of duplication here.

limonspb commented:
> Like @KarateBrot see a lot of duplication here.

Yes it will be some duplication at first. Now it's just dummy files for wings with empty functions

limonspb commented:
> > Like @KarateBrot see a lot of duplication here.
> 
> Yes it will be some duplication at first. Now it's just dummy files for wings with empty functions

Common code will be able to move in common files like autopilot.h when we know for sure what code is common )

SteveCEvans commented on src/main/pg/autopilot_wing.h:
```diff
@@ -0,0 +1,35 @@
+#include "pg/pg.h"
+
+typedef struct apConfig_s {
+} apConfig_t;
```
Wondering if an empty structure might cause any issues?

KarateBrot replied:
Good question. I looked it up and it is indeed undefined behavior. A pointer to an empty struct probably points to something else in the program with funny behavior "for free"

limonspb replied:
yeah just for safety i can fill it with `uint8 dummy`.

haslinghuis replied:
Yes it seems compiler dependent. See also https://gcc.gnu.org/onlinedocs/gcc-3.3.1/gcc/Empty-Structures.html#Empty%20Structures

@ledvinap what is your recommendation ?

KarateBrot commented:
I did a bad job at communicating my last comment, sorry. **This PR goes into the right direction**. It's a very good idea to separate the individual implementations.

What I think is "problematic" is controlling this kind of "static polymorphism" only via macro defines. Several files need to be kept in sync manually (function names between multiple independent files need to be kept the same) and there are no rules defined (no explicit interface), parameter groups can be completely different depending on macros, etc... which means code quality can go sideways really quickly if you don't know what you are doing (you obviously know what you are doing, I'm speaking in general terms, thinking about future contributors who want to extend the codebase). We can also not switch between quad and wing behavior in flight, for example for easy VTOLs. And adding just one another special implementation, for example if VTOL or something else needs its own file, that suddenly makes it so much more convoluted and complicated to manage all three.

At the same time, the current state of autopilot is challenging. It doesn't provide a nicely abstracted and generalized interface and that prevents you from going forward without first completely redoing huge parts of pos hold, alt hold and autopilot if you wanted to use a generalized interface. So keeping this challenge in mind, I think this PR is a nice work around or interim solution for the time being until we can eventually think of a good interface in autopilot.h that puts all special implementations together (without the dependence on build flags).

For example it would be nice if we can have wing and multirotor features built into one hex simultaneously if we want that. At the same time we could exclude multirotor features, if we only want wing features. Or we could exclude both, wing and multirotor. Doesn't make sense currently but that's a clean solution.

Refactoring or rather rewriting all the autonomous features is not something I look forward to but I guess it is something that needs to be done at some point in the future 🫠

haslinghuis commented on src/main/flight/pid.c:
```diff
@@ -570,7 +570,7 @@ STATIC_UNIT_TESTED FAST_CODE_NOINLINE float pidLevel(int axis, const pidProfile_
     angleTarget += gpsRescueAngle[axis] / 100.0f; // Angle is in centidegrees, stepped on roll at 10Hz but not on pitch
 #endif
-#ifdef USE_POSITION_HOLD
+#if defined(USE_POSITION_HOLD) && !defined(USE_WING)
     if (FLIGHT_MODE(POS_HOLD_MODE)) {
```
Perhaps we can use `isFixedWing()` instead of macro here
```suggestion
#if defined(USE_POSITION_HOLD)
    if (FLIGHT_MODE(POS_HOLD_MODE) && !isFixedWing()) {
```

limonspb replied:
If USE_WING is defined you can't use any of the autopilot even for drones now.
Once we have actual autopilot for wings, we will probably need to update isFixedWing function that it also checks the define.

Currently this line of code is correct.

limonspb commented:
@haslinghuis i've changed the lic header for the new files to a correct one. In the old files i left it as it was, just to make it clear its just a file renaming and adding #ifndef.
@SteveCEvans @KarateBrot i've added uint8_t dummy to the empty PG structures.
@ctzsnooze i've verified that this PR doesn't break pos hold, althold or GPS rescue. I don't have a mag, but was blown by how cool pos hold works even without it :) Little demo: https://www.youtube.com/watch?v=wTnCR-_wdkM

blckmn commented:
I'm still working through this one. I'm not sure there is specific functional differences between an auto pilot for wing versus quad. Vehicle in space being manoeuvred it's the control inputs (to position the vehicle) that vary not the auto pilot implementation itself.

Let me give this some more thought.

limonspb commented:
> I'm still working through this one. I'm not sure there is specific functional differences between an auto pilot for wing versus quad. Vehicle in space being manoeuvred it's the control inputs (to position the vehicle) that vary not the auto pilot implementation itself.
> 
> Let me give this some more thought.

The auto pilot interface (.h files) will be mostly the same in terms of "try go here" or "hold this altitude". But the implementation of it will have to be different in terms of how exactly it should be happening. For example for wings throttle controls the airspeed, not altitude. Roll controls the horizontal direction, pitch - altitude. It's also different PID controllers, and a different way of combining them into a single RPYT commands. Plus wings don't need the IMU-compass fusion, they rely just on GPS ground velocity.

Similar goes for GPS rescue - same interface as quads, but different stages, different behavior, different sanity checks, and different settings. In far further future it will also need separate complications like Downwind Leg, Base Leg, Final Approach for better landing.

blckmn commented:
But this is my point. Multi rotors can fly the pattern too for a landing. Autopilot is maintaining altitude, and heading (or a change in both). The control inputs will be different dependent on the vehicle but these are where implementation variance can occur... I think we risk separating too much of the code just because wing v multi... course tracking, establishing track error and correction required is the same.

limonspb commented:
> But this is my point. Multi rotors can fly the pattern too for a landing. Autopilot is maintaining altitude, and heading (or a change in both). The control inputs will be different dependent on the vehicle but these are where implementation variance can occur... I think we risk separating too much of the code just because wing v multi... course tracking, establishing track error and correction required is the same.

Not sure if we ever need quads to fly plane patterns for landing... Maybe in far future for V-toll?
There is a second risk im trying to avoid: combining the wing+multirotor code too much so it's much harder to work on, especially because it's different people working on them at the same time (now and probably later too). Would need lots of `ifs` and `ifdefs` covering many pieces of behavior and settings across all the files. I've done implementation for wings autopilot GPS rescue and it's mostly different code, except GPS sensor update. PG settings differ a lot too.

For example position control function for quads:
https://github.com/betaflight/betaflight/blob/master/src/main/flight/autopilot.c#L233
vs for wings (sorry, this wing code is in my branch, it works, but not super nice yet)
https://github.com/limonspb/betaflight/blob/NEW_WING_4.6_GPS/src/main/flight/gps_rescue_wing.c#L286

ctzsnooze commented:
I'm in favour of this solution while we are both working on the underlying code base for autopilot and how it would interact with GPS Rescue.

There are some fundamental considerations unresolved:
- exactly what kind of input requests should autopilot functions should be able to handle?
- should we be able to externally control the 'strength' or 'aggression' of the autopilot reactions

Until the 'client' code is written, the code that says, 'autopilot do this' or 'autopilot do that', we don't really know what we want autopilot to do, or what its inputs should be, or if they can be via some other function.

I experienced this a lot while writing the GPS Rescue update to use Autopilot's Position Control.  As a result, Position Control now has a function that shifts the target to a new point, enabling 'fly along a path'.  The method used is simple and works well with the existing Position Control methods.  

But perhaps we can make an alternative or better 'fly along a path' routine, within Autopilot code itself.

Hence a lot of this work is likely to remain 'fluid' while client applications are developed, each of which may have certain needs.  For instance, a 'geofence' or 'altitude limit' type function may require some optimisation of the position or altitude hold code - or may not - and we won't really know for sure until we go there.

At this point I think we should all be as 'open' as possible to making changes to all this stuff.

It is exceptionally difficult to test without flying it a lot; rescue and so on can't be done at the desk.  I know that both @limonspb and myself have spent hundreds of hours at the park doing nothing else.  I haven't flown a race quad now for almost two months, that's how involved it has been.  At the same time it has been very rewarding - we are definitely moving ahead. I congratulate Limon for how far he has progressed because I know how tough this stuff is.

I also think that it is really fantastic that Lemon is approaching this from a 'clean slate' perspective.

Both of us have attempted to implement the absolute simplest `method that works`, and by working absolutely independently, I think we have gained very different perspectives on things.

I had a long chat with @limonspb two days back.  We really thought hard about integration of wing and multi rotor autopilot, position hold, and rescue.

It became very clear to me that we cannot work on same files at the same time. We would be 'tripping over each others toes' all the time.  Worse, every time Limon made a change, we would have to test on both quad and wing, and some edge case issue may be added bit not be noticed until later.

Hence I think that while these functions are in active development, they should be completely independent.

We will of course work together and try to keep the code as similar as possible.

Once the code for each platform is functionally optimised, we can then look at whether we can reasonably simply include some wing-specific functions into a common file (eg just one autopilot file), or if we want to keep with very separate functions as per this PR.

For now I think that it is important to accept this PR so that both Limon and I can work in parallel.  Otherwise I think development for both mutlirotors and wings will be very much delayed.

haslinghuis reviewed (approved):
Perhaps we need to start somewhere to get the bigger picture for better abstractions.

SteveCEvans reviewed (approved):
My first thought was that it’s fine to split out the behaviours as we’d not need both quad and wing functionality in the same build, but then of course VTOL would require this.

I view this PR as an enabler to allow the quad and wing code to be developed separately, but we should ensure that a consistent interface is used by way of a table of function pointers to allow us to switch between implementations. @limonspb and @ctzsnooze  need to coordinate the evolution of this interface.

I’m therefore fine with this PR as is, but anticipate a future refactor and rationalisation as it’s determined what is common and what is different between quad and wing autopilot.
